Modules and firmware includes

hundeboll opened the pull request:
In order to include out-of-tree modules (and their required firmware) installed in an `updates` subfolder, a special case is added when a file is found below that folder.

Furthermore, a firmware file required by a module might be a symbolic link to the real firmware file. Add handling of this case too.

behrmann commented on mkosi/kmod.py:
```diff
@@ -234,12 +238,27 @@ def process_kernel_modules(
+                continue
+
+            if fw in required and (root / fw).is_symlink():
+                # mopdules may depend on a symbolic link to the actual firmware file, so
```
```suggestion
                # modules may depend on a symbolic link to the actual firmware file, so
```
there's also typos in the commit messages, `lnk` instead of `link` in the first and `an` instead of `and` in the second.

behrmann commented on mkosi/kmod.py:
```diff
@@ -218,12 +218,16 @@ def process_kernel_modules(
             firmware = sorted(firmwared.rglob("*"), reverse=True)
 
         for m in modules:
+            if m.is_relative_to(modulesd / "updates"):
+                continue
+
             if m in required:
```
This is a bit shorter and the comment below, would probably make more sense here. 
```suggestion
            if m in required or m.is_relative_to(modulesd / "updates"):
```
That being said, if files under `updates/` are required, that logic should be in `gen_required_kernel_modules`.

behrmann commented on mkosi/kmod.py:
```diff
@@ -234,12 +238,27 @@ def process_kernel_modules(
+            if fw.is_relative_to(firmwared / "updates"):
+                continue
+
+            if fw in required and (root / fw).is_symlink():
+                # mopdules may depend on a symbolic link to the actual firmware file, so
+                # make sure those target files are included in the final image too.
+                p = (root / fw).resolve().relative_to(root)
+                required.add(p)
+                unneeded.discard(p)
+                continue
+
             if fw in required:
                 continue
```
Since it tests the same thing again, this should probably be folded in, so that order between the blocks matters less.
```suggestion
            if fw in required
                if (root / fw).is_symlink():
                    # modules may depend on a symbolic link to the actual firmware file, so
                    # make sure those target files are included in the final image too.
                    p = (root / fw).resolve().relative_to(root)
                    required.add(p)
                    unneeded.discard(p)
                continue
```

DaanDeMeyer reviewed (changes requested):
This seems unnecessarily complicated, why not loop over `required`, calls `resolve()` on each entry and add all resolved paths to `required` as well?

hundeboll commented:
> This seems unnecessarily complicated, why not loop over `required`, calls `resolve()` on each entry and add all resolved paths to `required` as well?

Okay, spent some time to get to know `gen_required_kernel_modules()`. Better?

hundeboll commented:
Also removed the unconditional include of modules / firmwares below an `updates` folder, as the include / exlcude logic already handles those two. I've updated the documentation instead.

DaanDeMeyer commented on mkosi/kmod.py:
```diff
@@ -180,6 +180,11 @@ def gen_required_kernel_modules(
         firmware = set()
 
+    links = {
+        p.resolve().relative_to(root)
```
This will resolve the symlink relative to the host root directory which could lead to very wrong and very hard to debug issues depending on the host system.

Please use `chase()` from mkosi.sandbox to resolve the symlinks properly.

DaanDeMeyer commented on mkosi/kmod.py:
```diff
@@ -180,6 +180,11 @@ def gen_required_kernel_modules(
 
+    links = {
+        p.resolve().relative_to(root)
+        for p in filter(lambda fw: fw.is_symlink() and fw.exists(), map(lambda fw: root / fw, firmware))
```
Please use list comprehensions or a regular for loop, we don't use `filter()` or `map()` anywhere else, let's keep it that way for consistency.

DaanDeMeyer commented on mkosi/kmod.py:
```diff
@@ -189,6 +194,7 @@ def gen_required_kernel_modules(
             },
             mods,
             firmware,
+            links,
```
We should make sure the parent directories of these exist as well so please modify the first entry of the chain as well to make the parent directories for all these as well.

DaanDeMeyer commented on mkosi/kmod.py:
```diff
@@ -180,6 +180,11 @@ def gen_required_kernel_modules(
             mods = set(modulesd.rglob("*.ko*"))
         firmware = set()
 
+    links = {
```
This is a bad variable name, these aren't links, they're resolved links.

I don't think we need a new variable name for this, just adding the resolved symlinks to the `firmware` variable should be good enough.

This can also use a general comment explaining why we're doing this.

DaanDeMeyer commented on mkosi/kmod.py:
```diff
@@ -180,6 +181,14 @@ def gen_required_kernel_modules(
             mods = set(modulesd.rglob("*.ko*"))
         firmware = set()
 
+    # Some firmware dependencies are symbolic links, so the targets for those must be included in the list
+    # of required firmware files too.
+    for fw in firmware.copy():
+        if (root / fw).is_symlink():
+            target = Path(chase(root, fw))
+            if target.exists():
+                firmware.add(target.relative_to(root))
```
Maybe extend the comment and mention that we make an implicit assumption here that we only have a single level of symlinks. If we have a symlink to a symlink this approach won't work and we'll end up with a dangling symlink in the initrd.

DaanDeMeyer commented on mkosi/kmod.py:
```diff
@@ -180,6 +181,15 @@ def gen_required_kernel_modules(
         firmware = set()
 
+    # Some firmware dependencies are symbolic links, so the targets for those must be included in the list
+    # of required firmware files too. Links are not resolved recursively, and so links pointing to links
```
"links are not resolved recursively" => "intermediate symlinks are not included, ..."

behrmann commented on mkosi/resources/man/mkosi.1.md:
```diff
@@ -1016,10 +1016,11 @@ boolean argument: either `1`, `yes`, or `true` to enable, or `0`, `no`,
-    path (e.g. `i915` will match against `drivers/gpu/drm/i915.ko`). All modules that match any of the
-    specified patterns are included in the image. All module and firmware dependencies of the matched modules
-    are included in the image as well.
+    relative to the `/usr/lib/modules/<kver>/kernel` directory (note that they match against modules in the
+    `updates` and `extra` sibling directories too). mkosi checks for a match anywhere in the module path (e.g.
```
I think this would be clearer:
```suggestion
    relative to the `/usr/lib/modules/<kver>` directory. mkosi checks for a match anywhere in the module path (e.g.
```
mentioning the `updates` and `extra` directories explicitly only obscures that, we look relative to that.

behrmann commented on mkosi/kmod.py:
```diff
@@ -180,6 +181,15 @@ def gen_required_kernel_modules(
+    # results in dangling symlinks in the final image.
+    for fw in firmware.copy():
+        if (root / fw).is_symlink():
+            target = Path(chase(root, fw))
```
```suggestion
            target = Path(chase(os.fspath(root), os.fspath(fw)))
```
is needed here to appease mypy.

DaanDeMeyer commented on mkosi/kmod.py:
```diff
@@ -180,6 +181,15 @@ def gen_required_kernel_modules(
+    for fw in firmware.copy():
+        if (root / fw).is_symlink():
+            target = Path(chase(os.fspath(root), os.fspath(fw)))
+            if target.exists():
```
@hundeboll Is this check required? I don't think it's expected for the firmware directory to have dangling symlinks.

hundeboll replied:
Yes, `linux-firmware` on fedora rawhide has a dangling symlink which makes `cpio` fail later in the `mkosi` process:
```sh
# find /usr/lib/firmware/ -xtype l
/usr/lib/firmware/mrvl/sd8688_helper.bin.xz
```

DaanDeMeyer commented:
@hundeboll The ubuntu and centos initrd sizes need to be bumped by 1 in tests/test_initrd.py

septatrix commented on mkosi/kmod.py:
```diff
@@ -180,6 +181,15 @@ def gen_required_kernel_modules(
+    # Some firmware dependencies are symbolic links, so the targets for those must be included in the list
+    # of required firmware files too. Intermediate symlinks are not included, and so links pointing to links
+    # results in dangling symlinks in the final image.
+    for fw in firmware.copy():
```
Is the call to `copy` even required here?

behrmann replied:
Otherwise this is mutating the set while it is being iterated over, which throws a `RuntimeError: Set changed size during iteration`.
